pkg/reconciler/tenancy/initialization/apibinder_initializer_reconcile.go

```diff
@@ -0,0 +1,280 @@
+	// were errors trying to initialize APIBindings for the workspace.
+	WorkspaceInitializedAPIBindingErrors = "APIBindingErrors"
+)
+
+func (b *APIBinder) reconcile(ctx context.Context, clusterWorkspace *tenancyv1alpha1.ClusterWorkspace) error {
+	logger := klog.FromContext(ctx).WithValues(
+		"clusterWorkspaceType.path", clusterWorkspace.Spec.Type.Path,
+		"clusterWorkspaceType.name", clusterWorkspace.Spec.Type.Name,
+	)
+
+	var errors []error
+	clusterName := logicalcluster.From(clusterWorkspace).Join(clusterWorkspace.Name)
+	logger.V(2).Info("initializing APIBindings for workspace")
+
+	// Start with the ClusterWorkspaceType specified by the ClusterWorkspace
+	leafCWT, err := b.getClusterWorkspaceType(logicalcluster.New(clusterWorkspace.Spec.Type.Path), string(clusterWorkspace.Spec.Type.Name))
+	if err != nil {
+		logger.Error(err, "error getting ClusterWorkspaceType")
+
+		conditions.MarkFalse(
+			clusterWorkspace,
+			WorkspaceAPIBindingsInitialized,
+			WorkspaceInitializedClusterWorkspaceTypeInvalid,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error getting ClusterWorkspaceType %s|%s: %v",
+			clusterWorkspace.Spec.Type.Path, clusterWorkspace.Spec.Type.Name,
+			err,
+		)
+
+		return nil
+	}
+
+	// Get all the transitive ClusterWorkspaceTypes
+	cwts, err := b.transitiveTypeResolver.Resolve(leafCWT)
+	if err != nil {
+		logger.Error(err, "error resolving transitive types")
+
+		conditions.MarkFalse(
+			clusterWorkspace,
+			WorkspaceAPIBindingsInitialized,
+			WorkspaceInitializedClusterWorkspaceTypeInvalid,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error resolving transitive set of cluster workspace types: %v",
+			err,
+		)
+
+		return nil
+	}
+
+	// Get current bindings
+	bindings, err := b.listAPIBindings(clusterName)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	// This keeps track of which APIBindings have been created for which APIExports
+	exportToBinding := map[apisv1alpha1.WorkspaceExportReference]*apisv1alpha1.APIBinding{}
+
+	for i := range bindings {
+		binding := bindings[i]
+
+		if binding.Spec.Reference.Workspace == nil {
+			continue
+		}
+
+		// Track what we have ("actual")
+		exportToBinding[*binding.Spec.Reference.Workspace] = binding
+	}
+
+	requiredExportRefs := map[tenancyv1alpha1.APIExportReference]struct{}{}
+
+	for _, cwt := range cwts {
+		logger := logging.WithObject(logger, cwt)
+		logger.V(2).Info("attempting to initialize APIBindings")
+
+		for i := range cwt.Spec.DefaultAPIBindings {
+			exportRef := cwt.Spec.DefaultAPIBindings[i]
+
+			// Keep track of unique set of expected exports across all CWTs
+			requiredExportRefs[exportRef] = struct{}{}
+
+			logger := logger.WithValues("apiExport.path", exportRef.Path, "apiExport.name", exportRef.Name)
+			ctx := klog.NewContext(ctx, logger)
+
+			// Make sure the APIExport exists
+			apiExport, err := b.getAPIExport(logicalcluster.New(exportRef.Path), exportRef.Name)
```

Comment:
won't a missing export cause the binding create call to fail? why is the separate check needed? are we going through the VW in this get, or will we leak info the user may not have permission to see?

Reply:
We don't have admission that rejects APIBindings if the APIExport is invalid (maybe we should?).

This is a cached read, not through the VW.

I don't think this will leak anything - you'll be able to see the APIBinding in the workspace, and see what APIExport it's referencing.

Reply:
Hm, I thought that the lack of the workspace or the export would today lead to that "missing verb='bind'" error (but in general once that's fixed would give the appropriate error - whereas here we use super-user credentials to a) check for the workspace existence and b) check for the export existence, neither of which the user should be able to see?

Reply:
You're right that RBAC has to be in place to be able to create an APIBinding. If we fix #1939, I guess there's a chance someone could have permission to bind to `*` resourceNames, in which case you could create a reference to a non-existent export?

> check for the workspace existence

Where is that?

> check for the export existence, neither of which the user should be able to see?

Given that the APIBinding spec contains a reference to an APIExport by path+name, I'm not sure that we're leaking anything here, right?

Reply:
Just create 1000x APIExportReferences and go fishing to find exports that exist?

Reply:
I will remove it if you think that's better

Reply:
I'm not a security expert but to me this looks like an unnecessary leak - happy to check with some other folks to clarify

Reply:
I am currently also retrieving the APIExport to get its permission claims to auto-accept them. I can

1. keep everything as-is
2. keep the current code to get the export, but if we get an error there, translate the error to something less leaky
3. don't get the export, and instead create the binding w/o accepted permission claims
    1. once the binding has been created, wait for permission claims to show up in status, then use those to accept (how do we distinguish between 0 claims because 0, vs. 0 claims because reconciliation hasn't happened yet?)